armory/utils/metrics.py

```diff
@@ -626,7 +627,181 @@ def object_detection_AP_per_class(list_of_ys, list_of_y_preds):
+
+    The box associated with the patch is assigned the label of the patch's targeted class.
+    Thus, a true positive is the case where the detector predicts the patch's targeted
+    class (at a location overlapping the patch). A false positive is the case where the
+    detector predicts a non-targeted class at a location overlapping the patch. If the
+    detector predicts multiple instances of the target class (that overlap with the patch),
+    one of the predictions is considered a true positive and the others are ignored.
+
+    This metric is computed over all evaluation samples, rather than on a per-sample basis.
+    It returns a dictionary mapping each class to the average precision (AP) for the class.
+    The only classes with potentially nonzero AP's are the classes targeted by the patches
+    (see above paragraph).
+    """
+
+    # From https://arxiv.org/abs/1912.08166: use a low IOU since "the patches will sometimes
+    # generate many small, overlapping predictions in the region of the attack"
+    IOU_THRESHOLD = 0.1
+
+    # Precision will be computed at recall points of 0, 0.1, 0.2, ..., 1
+    RECALL_POINTS = np.linspace(0, 1, 11)
+
+    # Converting boxes to a list of dicts (a list for predicted boxes that overlap with the patch,
+    # and a separate list for ground truth patch boxes), where each dict corresponds to a box and
+    # has the following keys "img_idx", "label", "box", as well as "score" for predicted boxes
+    patch_boxes_list = []
+    overlappping_pred_boxes_list = []
+    for img_idx, (y, y_pred) in enumerate(zip(list_of_ys, list_of_y_preds)):
+        idx_of_patch = np.where(y["labels"] == ADV_PATCH_MAGIC_NUMBER_LABEL_ID)
+        patch_box = y["boxes"][idx_of_patch].flatten()
+        patch_id = y["patch_id"][idx_of_patch]
+        patch_target_label = [
+            patch["adv_target"] for patch in APRICOT_PATCHES if patch["id"] == patch_id
+        ][0]
+        patch_box_dict = {
+            "img_idx": img_idx,
+            "label": patch_target_label,
+            "box": patch_box,
+        }
+        patch_boxes_list.append(patch_box_dict)
+
+        for pred_box_idx in range(len(y_pred["labels"].flatten())):
+            box = y_pred["boxes"][pred_box_idx]
+            if _intersection_over_union(box, patch_box) > IOU_THRESHOLD:
+                label = y_pred["labels"][pred_box_idx]
+                score = y_pred["scores"][pred_box_idx]
+                pred_box_dict = {
+                    "img_idx": img_idx,
+                    "label": label,
+                    "box": box,
+                    "score": score,
+                }
+                overlappping_pred_boxes_list.append(pred_box_dict)
+
+    # Union of (1) the set of classes targeted by patches and (2) the set of all classes
+    # predicted at a location that overlaps the patch in the image
+    set_of_class_ids = set([i["label"] for i in patch_boxes_list]) | set(
```

Comment:
I had considered this, but opted for duplicating some code in the interest of the code's clarity. Beyond multiple matches to the same box, the functions also differ in that for this metric you don't need to check the iou since the predictions are already filtered by iou with the patch. The differences in the variable names and comments helps make it clear what this metric is computing. A helper function, I think, would get a little bit convoluted having to handle "regular mAP" mode as well as "apricot" mode